File: app/components/lookbook/elements/tab_group/tab_group.rb
module Lookbook
  class TabGroup < Lookbook::BaseComponent
    include AcceptsOptionsConcern

    accepts_option :orientation, from: [:horizontal, :vertical], default: :horizontal

    with_slot :tab do |name = nil, **kwargs|
      kwargs = kwargs.deep_merge({
        x: {init: "registerTab"},
        orientation: get_option_value(:orientation)
      })

      Lookbook::Tab.new(name: name, **kwargs)
    end

    def initialize(name:, **kwargs)
      raise "Missing tab group name" unless name.present?
      set_tag_data_attr("tab-group-id", name)
    end
  end
end

File: app/components/lookbook/elements/tab_group/tab_group.js
import { getComponent } from "@helpers/alpine";
import Component from "@js/component";

export default Component("tabGroup", () => {
  return {
    $options: {},
    options: ["orientation"],

    groupId: null,
    tabs: [],
    selectedTab: null,

    init() {
      this.groupId = this.$el.getAttribute("data-tab-group-id");
      this.$nextTick(() => {
        this.tabs[0].selectTab();
      });
    },

    registerTab() {
      const tab = getComponent(this.$el);
      this.tabs.push(tab);
    },

    selectTab() {
      this.selectedTab = this.id;
      this.$dispatch("tab:selected", {
        group: this.groupId,
        tab: this.selectedTab,
      });
    },
  };
});
